src/DotNetty.Handlers/Timeout/WriteTimeoutHandler.cs

```diff
@@ -0,0 +1,199 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using DotNetty.Common.Internal;
+
+namespace DotNetty.Handlers.Timeout
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using DotNetty.Common.Utilities;
+    using DotNetty.Common.Concurrency;
+    using DotNetty.Transport.Channels;
+
+    /// <summary>
+    /// Raises a <see cref="WriteTimeoutException"/> when a write operation cannot finish in a certain period of time.
+    /// 
+    /// <para>
+    /// <example>
+    /// 
+    /// The connection is closed when a write operation cannot finish in 30 seconds.
+    ///
+    /// <c>
+    /// var bootstrap = new <see cref="DotNetty.Transport.Bootstrapping.ServerBootstrap"/>();
+    ///
+    /// bootstrap.ChildHandler(new ActionChannelInitializer<ISocketChannel>(channel =>
+    /// {
+    ///     IChannelPipeline pipeline = channel.Pipeline;
+    ///     
+    ///     pipeline.AddLast("writeTimeoutHandler", new <see cref="WriteTimeoutHandler"/>(30);
+    ///     pipeline.AddLast("myHandler", new MyHandler());
+    /// }    
+    /// </c>
+    /// 
+    /// <c>
+    /// public class MyHandler : ChannelDuplexHandler 
+    /// {
+    ///     public override void ExceptionCaught(<see cref="IChannelHandlerContext"/> context, <see cref="Exception"/> exception)
+    ///     {
+    ///         if(exception is <see cref="WriteTimeoutException"/>) 
+    ///         {
+    ///             // do somethind
+    ///         }
+    ///         else
+    ///         {
+    ///             base.ExceptionCaught(context, cause);
+    ///         }
+    ///      }
+    /// }
+    /// </c>
+    /// 
+    /// </para>
+    /// <see cref="ReadTimeoutHandler"/>
+    /// <see cref="IdleStateHandler"/>
+    /// </summary>
+    public class WriteTimeoutHandler : ChannelHandlerAdapter
+    {
+        static readonly TimeSpan MinTimeout = TimeSpan.FromMilliseconds(1);
+        readonly TimeSpan timeout;
+        bool closed;
+
+        /// <summary>
+        /// A doubly-linked list to track all WriteTimeoutTasks.
+        /// </summary>
+        readonly LinkedList<WriteTimeoutTask> tasks = new LinkedList<WriteTimeoutTask>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.ReadTimeoutHandler"/> class.
+        /// </summary>
+        /// <param name="timeoutSeconds">Timeout in seconds.</param>
+        public WriteTimeoutHandler(int timeoutSeconds)
+            : this(TimeSpan.FromSeconds(timeoutSeconds))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DotNetty.Handlers.Timeout.ReadTimeoutHandler"/> class.
+        /// </summary>
+        /// <param name="timeout">Timeout.</param>
+        public WriteTimeoutHandler(TimeSpan timeout)
+        {
+            this.timeout = 
+                timeout.Ticks <= 0 ? MinTimeout : TimeUtil.Max(timeout, MinTimeout);
```

Comment:
is `<= 0` check necessary here? it seems `.Max(..)` will take care of that as well.